nixos/modules/services/security/oauth2_proxy.nix

```diff
@@ -0,0 +1,528 @@
+# NixOS module for oauth2_proxy.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.oauth2_proxy;
+
+  # Use like:
+  #   repeatedArgs (arg: "--arg=${arg}") args
+  repeatedArgs = concatMapStringsSep " ";
+
+  # 'toString' doesn't quite do what we want for bools.
+  fromBool = x: if x then "true" else "false";
+
+  # oauth2_proxy provides many options that are only relevant if you are using
+  # a certain provider. This set maps from provider name to a function that
+  # takes the configuration and returns a string that can be inserted into the
+  # command-line to launch oauth2_proxy.
+  providerSpecificOptions = {
+    azure = cfg: ''
+      --azure-tenant=${cfg.azure.tenant} \
+      --resource=${cfg.azure.resource} \
+    '';
+
+    github = cfg: ''
+      $(optionalString (!isNull cfg.github.org) "--github-org=${cfg.github.org}") \
+      $(optionalString (!isNull cfg.github.team) "--github-org=${cfg.github.team}") \
+    '';
+
+    google = cfg: ''
+      --google-admin-email=${cfg.google.adminEmail} \
+      --google-service-account=${cfg.google.serviceAccountJSON} \
+      $(repeatedArgs (group: "--google-group=${group}") cfg.google.groups) \
+    '';
+  };
+
+  authenticatedEmailsFile = pkgs.writeText "authenticated-emails" cfg.email.addresses;
+
+  getProviderOptions = cfg: provider:
+    if providerSpecificOptions ? provider then providerSpecificOptions.provider cfg else "";
+
+  mkCommandLine = cfg: ''
+    --provider=${cfg.provider} \
+    ${optionalString (!isNull cfg.email.addresses) "--authenticated-emails-file=${authenticatedEmailsFile}"} \
+    --approval-prompt=${cfg.approvalPrompt} \
+    ${optionalString (cfg.passBasicAuth && !isNull cfg.basicAuthPassword) "--basic-auth-password=${cfg.basicAuthPassword}"} \
+    --client-id=${cfg.clientID} \
+    --client-secret=${cfg.clientSecret} \
+    ${optionalString (!isNull cfg.cookie.domain) "--cookie-domain=${cfg.cookie.domain}"} \
+    --cookie-expire=${cfg.cookie.expire} \
+    --cookie-httponly=${fromBool cfg.cookie.httpOnly} \
+    --cookie-name=${cfg.cookie.name} \
+    --cookie-secret=${cfg.cookie.secret} \
```

Comment:
Hey, first: nice work.

I just tried this and I think the values need either explicit escaping or at least `'..'` around the arguments

[1]
https://www.freedesktop.org/software/systemd/man/systemd-escape.html